### PR TITLE
Add `strict_settings` option, allow runtime fallbacks for custom settings

### LINKED ISSUE #417: Incompatible with django-configurations due to the way the settings are imported
# Bug report
:bug: 

## What's wrong

Trying to use `django-stubs` together with `django-configurations` is a big pain since the latter interferes and controls how django settings are loaded. By default, the following exception gets thrown when trying to run `mypy`:
```bash
(mypackage-vfeg9Sk1-py3.6) ➜  mypackage IN-3486 ✗ python -m mypy -p src.mypackage                                                       
Error constructing plugin instance of NewSemanalDjangoPlugin

Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/mypy/__main__.py", line 12, in <module>
    main(None, sys.stdout, sys.stderr)
  File "mypy/main.py", line 89, in main
  File "mypy/build.py", line 180, in build
  File "mypy/build.py", line 227, in _build
  File "mypy/build.py", line 465, in load_plugins
  File "mypy/build.py", line 443, in load_plugins_from_config
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/mypy_django_plugin/main.py", line 80, in __init__
    self.django_context = DjangoContext(django_settings_module)
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/mypy_django_plugin/django/context.py", line 88, in __init__
    apps, settings = initialize_django(self.django_settings_module)
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/mypy_django_plugin/django/context.py", line 70, in initialize_django
    settings._setup()
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/django/conf/__init__.py", line 66, in _setup
    self._wrapped = Settings(settings_module)
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/django/conf/__init__.py", line 157, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/sarunas/Documents/projects/mypackage/mypackage/project/settings.py", line 6, in <module>
    from mypackage.settings import Base, IntegratedTests
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/mypackage/settings.py", line 26, in <module>
    class Base(Configuration):  # pylint: disable=no-init
  File "/home/sarunas/.cache/pypoetry/virtualenvs/mypackage-vfeg9Sk1-py3.6/lib/python3.6/site-packages/configurations/base.py", line 28, in __new__
    raise ImproperlyConfigured(install_failure)
django.core.exceptions.ImproperlyConfigured: django-configurations settings importer wasn't correctly installed. Please use one of the starter functions to install it as mentioned in the docs: https://django-configurations.readthedocs.io/

```
That makes sense, since `configurations` disallows loading the settings manually and requires to run these two calls to configure django, both of them are for example found in `manage.py` and `wsgi.py`:
```python
        import configurations
        configurations.setup()
```

**mypy configuration in setup.cfg** (it follows `[metadata]` and `[options]` package metadata definitions)
```ini
[mypy.plugins.django-stubs]
django_settings_module = project.settings

[mypy]
plugins =
    mypy_django_plugin.main

ignore_errors =            false
warn_no_return =           true
strict_equality =          true
strict_optional =          true
warn_return_any =          true
warn_unreachable =         true
implicit_reexport =        false
check_untyped_defs =       true
local_partial_types =      true
warn_unused_ignores =      true
warn_unused_configs =      true
warn_redundant_casts =     true
no_implicit_optional =     true

namespace_packages =       true
fine_grained_incremental = true
show_error_codes =         true

strict =                   true
allow_any_generics =       false
allow_subclassing_any =    true
ignore_missing_imports =   false

[mypy-mypackage.migrations.*]
ignore_errors = true
```

I tried looking for a workaround which didn't involve changing neither `mypy` nor `django-stubs`, nor `django-configurations` source files and I'm afraid I have failed. Unfortunately this also blocks DRF stubs from being useful.

## How is that should be
`django-configurations` should be supported and allowed to load django settings if the user uses it (unless it is somehow incompatible with the way the stubs are generated?). 

## System information

- OS: "Ubuntu 18.04.4 LTS"
- `python` version: `3.6.9`
- `django` version: `2.2.14`
- `mypy` version: `0.770`
- `django-stubs` version: `1.5.0`
- `django-configurations` version: `2.2`

## Possible solution
One of possible options would be to include those two calls in `mypy_django_plugin.django.context` `initialize_django` call, just before the `models` get imported:

```python
def initialize_django(settings_module: str) -> Tuple['Apps', 'LazySettings']:
    with temp_environ():
        os.environ['DJANGO_SETTINGS_MODULE'] = settings_module

        # add current directory to sys.path
        sys.path.append(os.getcwd())

        def noop_class_getitem(cls, key):
            return cls
        
        try:
             # or use importlib.import_module, or check whether django uses it
            import configurations
            configurations.setup()
        except ModuleNotFoundError:
            pass

        from django.db import models

        models.QuerySet.__class_getitem__ = classmethod(noop_class_getitem)  # type: ignore
        models.Manager.__class_getitem__ = classmethod(noop_class_getitem)  # type: ignore
        ...
```

I found that this works for me - though I haven't tested it extensively yet. I'd be happy to add a PR for this upon your confirmation.

### LINKED ISSUE #1276: Dynamic settings fail as of 1.13.0
# Bug report

## What's wrong

We build Django settings dynamically, instead of using a `settings.py` file with static properties.  This worked fine with django-stubs up until 1.13.0, but all of our settings fail with 1.13.0:

```
'Settings' object has no attribute 'FOO'
```

This seems to be caused by #1163.

## How is that should be

I understand the reasoning given in #1163, but we are pinned at django-stubs 1.12.0 and would have to completely reimplement our installation in order to unblock.  Can dynamically-constructed settings be supported with django-stubs, or is that door permanently closed?  Maybe it could be a non-default option we could enable, accepting the potential downsides?

## System information

- OS:
- `python` version: 3.10.8
- `django` version: 4.1.3
- `mypy` version: 0.991
- `django-stubs` version: 1.13.0
- `django-stubs-ext` version: 0.7.0

### LINKED ISSUE #1410: Custom django settings generate mypy errors
## What's wrong

We have a project that is based on the wemake template. We created the project years ago some stuff might have changed with respect to the latest template

We have added some custom settings to server/settings/components/common.py. :
```
# common.py
MIN_GH_VERSION = "1.0.0"
```

Since we updated to latest version of django-stubs, we get mypy errors when we reference these errors in other parts of the code:

```
server/runner/logic/versioning.py:29: error: 'Settings' object has no attribute 'MIN_GH_VERSION'
```` 

We come from django-stubs 1.12.0 and are now on 1.16.0. We are on mypy version 1.1.1

django-stubs in setup.cfg:
```
[mypy.plugins.django-stubs]
django_settings_module = server.settings
```

- OS:
- `python` version: 3.10
- `django` version: 3.2.18
- `mypy` version: 1.1.1
- `django-stubs` version: 1.16.0
- `django-stubs-ext` version: 0.8.0

edit: traced it back to this pull request I think: https://github.com/typeddjango/django-stubs/pull/1175

### PR DESCRIPTION
**Pull request by sobolevn**
This PR adds a new setting to our plugin called `strict_settings`.
When set to `True` (the default), it raises for any statically unknown setting value.

But, when set to `False`, it checks for the runtime setting value and if it is present, it returns `Any` in all cases.

This will change will make it possible to use `django-stubs` with [`django-split-settings`](https://github.com/wemake-services/django-split-settings) (which I maintain and use for all my projects) and similar tools like `django-configurations`.

After https://github.com/typeddjango/django-stubs/pull/1163 it was very hard to use settings in dynamic workflows.

Closes https://github.com/typeddjango/django-stubs/issues/1410
Closes https://github.com/typeddjango/django-stubs/issues/1276
Closes https://github.com/typeddjango/django-stubs/issues/417

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -311,6 +311,52 @@ If you encounter this error in your own code, you can either cast the `Promise`
...
+So, mypy would not like this code:
+
+```python
+from django.conf import settings
+
+settings.CUSTOM_VALUE  # E: 'Settings' object has no attribute 'CUSTOM_SETTING'
+```
+
+To handle this corner case we have a special setting `strict_settings` (`True` by default),
+you can switch it to `False` to always return `Any` and not raise any errors if runtime settings module has the given value:
```
Comment (intgr) [inline README.md]: Could use clarifying, which example is for ini and which for toml.

```suggestion
you can switch it to `False` to always return `Any` and not raise any errors if runtime settings module has the given value,
for example `pyproject.toml`:
```

File: README.md [hunk 2]
```diff
@@ -311,6 +311,52 @@ If you encounter this error in your own code, you can either cast the `Promise`
...
+
+```toml
+[tool.django-stubs]
+strict_settings = false
+```
+
+or
```
Comment (intgr) [inline README.md]:
```suggestion
or `mypy.ini`:
```

File: tests/typecheck/test_settings.yml [hunk 3]
```diff
@@ -58,3 +58,32 @@
...
       main:5: error: 'Settings' object has no attribute 'NON_EXISTANT_SETTING'
       main:5: note: Revealed type is "Any"
+
+
+-   case: settings_loaded_from_runtime_magic
+    disable_cache: true
+    main: |
+        from django.conf import settings
+        reveal_type(settings.A)  # N: Revealed type is "Any"
+        reveal_type(settings.B)  # E: 'Settings' object has no attribute 'B' # N: Revealed type is "Any"
```
Comment (intgr) [inline tests/typecheck/test_settings.yml:73]: What about Django builtin settings like `settings.SECRET_KEY`, I think those should still have concrete types?

File: README.md [hunk 4]
```diff
@@ -311,6 +311,52 @@ If you encounter this error in your own code, you can either cast the `Promise`
...
+
+```ini
+[mypy.plugins.django-stubs]
+strict_settings = false
+```
```
Comment (intgr) [inline README.md:349]: Although there are just 2 settings for now, I would prefer it if all django-stubs settings were documented in one place. Maybe add another README section for settings and then link to it from here?

File: mypy_django_plugin/config.py [hunk 5]
```diff
@@ -14,21 +14,23 @@
...
+strict_settings = bool (default: true)
 ...
 """
 INVALID_FILE = "mypy config file is not specified or found"
 COULD_NOT_LOAD_FILE = "could not load configuration file"
-MISSING_SECTION = "no section [{section}] found".format
+MISSING_SECTION = "no section [{section}] found"
 MISSING_DJANGO_SETTINGS = "missing required 'django_settings_module' config"
-INVALID_SETTING = "invalid {key!r}: the setting must be a boolean".format
+INVALID_BOOL_SETTING = "invalid {key!r}: the setting must be a boolean"
```
Comment (intgr) [inline mypy_django_plugin/config.py:33]: This was weird, thanks for cleaning up.

File: tests/typecheck/test_settings.yml (hunk 3 above)
Comment (sobolevn) [reply tests/typecheck/test_settings.yml:73]: They are checked above, this setting does not change the default behaviour.

File: README.md (hunk 4 above)
Comment (sobolevn) [reply README.md:349]: Yes, we can do this later. Good idea 👍

File: tests/typecheck/test_settings.yml (hunk 3 above)
Comment (intgr) [reply tests/typecheck/test_settings.yml:73]: Let's add one such setting to this `strict_settings = false` test case too.

**Comment by intgr** [comment]
When I try to run this in one of my projects, I get the following error. Haven't investigated at all.

```
mypy 1.4.0 (compiled: yes)
Traceback (most recent call last):
  File ".../bin/mypy", line 8, in <module>
Error constructing plugin instance of NewSemanalDRFPlugin

    sys.exit(console_entry())
             ^^^^^^^^^^^^^^^
  File ".../lib/python3.11/site-packages/mypy/__main__.py", line 15, in console_entry
    main()
  File "mypy/main.py", line 95, in main
  ...
  File "mypy/build.py", line 484, in load_plugins_from_config
  File ".../lib/python3.11/site-packages/mypy_drf_plugin/main.py", line 26, in __init__
    self.django_context = DjangoContext(django_settings_module)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.11/site-packages/mypy_django_plugin/django/context.py", line 85, in __init__
    apps, settings = initialize_django(self.plugin_config.django_settings_module)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'django_settings_module'
```

**Comment by intgr** [comment]
Looks like some interaction between `djangorestframework-stubs` and `django-stubs`.

**Comment by sobolevn** [comment]
Oh, good catch! I've change the signature of `DjangoContext`, so we would need to change it here as well: https://github.com/typeddjango/djangorestframework-stubs/blob/8709db5d2af980919a46072725ab54ca637a8703/mypy_drf_plugin/main.py#L25-L26

I've opened https://github.com/typeddjango/djangorestframework-stubs/issues/432

**Comment by UnknownPlatypus** [comment]
If I understand well, this makes it possible to use django-stubs along `django-configuration` or `django-split-settings` but the types inferred will be `Any` right ?

To have more accurate typing, one should make his own plugin ?

**Comment by sobolevn** [comment]
> If I understand well, this makes it possible to use django-stubs along django-configuration or django-split-settings but the types inferred will be Any right ?

Yes!

> To have more accurate typing, one should make his own plugin ?

I can work on this later on, I need a way to infer mypy types from runtime types. I have a prototype that does just this. But, I need to polish it for a bit more.

File: mypy_django_plugin/django/context.py [hunk 6]
```diff
@@ -71,13 +72,20 @@ class LookupsAreUnsupported(Exception):
 
 
 class DjangoContext:
-    def __init__(self, django_settings_module: str) -> None:
-        self.django_settings_module = django_settings_module
+    def __init__(self, plugin_config: DjangoPluginConfig) -> None:
+        if isinstance(plugin_config, str):
+            raise RuntimeError("Old API is used: we now require DjangoPluginConfig instance")
```
Comment (intgr) [inline mypy_django_plugin/django/context.py]: So people who have older djangorestframework-stubs installed and update django-stubs will receive this message?

Seems a bit cryptic. How about "Older djangorestframework-stubs versions are no longer supported, please update djangorestframework-stubs" or similar?

I doubt there are any other projects that import djagno-stubs internals, anyway.

Comment (sobolevn) [reply mypy_django_plugin/django/context.py]: The only project is `drf-stubs`. I don't really like this solution now. See https://github.com/typeddjango/djangorestframework-stubs/pull/433

I propose this plan:
- [x] I remove this error
- [ ] We release a new version of `drf-stubs` with the upper bound of `django-stubs` set to the current version of `django-stubs` (`>=4.2,<=4.2.1`)
- [ ] We merge this and release a new version of `django-stubs`
- [ ] We merge https://github.com/typeddjango/djangorestframework-stubs/pull/433 and release a new version of `drf-stubs` with no upper bound

Comment (sobolevn) [reply mypy_django_plugin/django/context.py]: On the other hand, it won't solve the problem of only upgrading `django-stubs` with old `drf-stubs` 😞 

Ideas?

Comment (sobolevn) [reply mypy_django_plugin/django/context.py]: Ok, I was able to solve all the compat problems in Python code. The easiest way for some reason is always the hardest one to find.

**Comment by sobolevn** [comment]
I want to merge this today, so I can work on https://github.com/typeddjango/django-stubs/issues/1568

@intgr any other comments? :)

**Review by intgr — COMMENTED** [review]
Sorry, I was away for a few days. But looks good to me now.

**Comment by sobolevn** [comment]
No worries, happy to see that you like my fix :)

**Comment by kaldown** [comment]
This do not work with `django-configurations`.

```
[tool.mypy]
mypy_path = "./src"
plugins = ["mypy_django_plugin.main"]

[tool.django-stubs]
django_settings_module = "myproj.settings"
strict_settings = false
```

returns an Error:
```
Error constructing plugin instance of NewSemanalDjangoPlugin
...
raise ImproperlyConfigured(install_failure)
django.core.exceptions.ImproperlyConfigured: django-configurations settings importer wasn't correctly installed. Please use one of the starter functions to install it as mentioned in the docs: https://django-configurations.readthedocs.io/
```

Whilst commenting usage of **django-stubs** in **[tool.mypy]** directive of **pyproject.toml** leading to expected issue of **mypy** not being able to catch _django_ types from _models.py_ (for example).

As I see this PR https://github.com/typeddjango/django-stubs/pull/180 is still open.
Would it be better to remove **django-configurations** entry from _README.md_ since it could cause confusion?

**Comment by intgr** [comment]
@kaldown Please open a separate issue about this.

**Comment by kaldown** [comment]
@intgr thank you, for the fast reply. Do you think it's better to open another one in addition to https://github.com/typeddjango/django-stubs/pull/180?

**Comment by intgr** [comment]
Yes.

**Comment by kaldown** [comment]
Done: https://github.com/typeddjango/django-stubs/issues/1709

**Pull request merged**